pubsublite/integration_test.go

```diff
@@ -0,0 +1,271 @@
+	region, _ := ZoneToRegion(zone)
+	resourceID := fmt.Sprintf("go-test-admin-%d", rng.Int63())
+
+	locationPath := LocationPath{Project: proj, Zone: zone}
+	topicPath := TopicPath{Project: proj, Zone: zone, TopicID: resourceID}
+	subscriptionPath := SubscriptionPath{Project: proj, Zone: zone, SubscriptionID: resourceID}
+	t.Logf("Topic path: %s", topicPath)
+
+	client := adminClient(ctx, t, region)
+	defer client.Close()
+
+	// Topic admin operations.
+	newTopicConfig := &TopicConfig{
+		Name:                       topicPath,
+		PartitionCount:             2,
+		PublishCapacityMiBPerSec:   4,
+		SubscribeCapacityMiBPerSec: 4,
+		PerPartitionBytes:          30 * gibi,
+		RetentionDuration:          time.Duration(24 * time.Hour),
+	}
+
+	gotTopicConfig, err := client.CreateTopic(ctx, newTopicConfig)
+	if err != nil {
+		t.Fatalf("Failed to create topic: %v", err)
+	}
+	defer cleanUpTopic(ctx, t, client, topicPath)
+	if !testutil.Equal(gotTopicConfig, newTopicConfig) {
+		t.Errorf("CreateTopic() got: %v, want: %v", gotTopicConfig, newTopicConfig)
+	}
+
+	if gotTopicConfig, err := client.Topic(ctx, topicPath); err != nil {
+		t.Errorf("Failed to get topic: %v", err)
+	} else if !testutil.Equal(gotTopicConfig, newTopicConfig) {
+		t.Errorf("Topic() got: %v, want: %v", gotTopicConfig, newTopicConfig)
+	}
+
+	if gotTopicPartitions, err := client.TopicPartitions(ctx, topicPath); err != nil {
+		t.Errorf("Failed to get topic partitions: %v", err)
+	} else if gotTopicPartitions != newTopicConfig.PartitionCount {
+		t.Errorf("TopicPartitions() got: %v, want: %v", gotTopicPartitions, newTopicConfig.PartitionCount)
+	}
+
+	topicIt := client.Topics(ctx, locationPath)
+	var foundTopic *TopicConfig
+	for {
+		topic, err := topicIt.Next()
+		if err == iterator.Done {
+			break
+		}
+		if testutil.Equal(topic.Name, topicPath) {
+			foundTopic = topic
+			break
+		}
+	}
+	if foundTopic == nil {
+		t.Error("Topics() did not return topic config")
+	} else if !testutil.Equal(foundTopic, newTopicConfig) {
+		t.Errorf("Topics() found config: %v, want: %v", foundTopic, newTopicConfig)
+	}
+
+	topicUpdate1 := &TopicConfigToUpdate{
+		Name:                       topicPath,
+		PublishCapacityMiBPerSec:   6,
+		SubscribeCapacityMiBPerSec: 8,
+	}
+	wantUpdatedTopicConfig1 := &TopicConfig{
+		Name:                       topicPath,
+		PartitionCount:             2,
+		PublishCapacityMiBPerSec:   6,
+		SubscribeCapacityMiBPerSec: 8,
+		PerPartitionBytes:          30 * gibi,
+		RetentionDuration:          time.Duration(24 * time.Hour),
+	}
+	if gotTopicConfig, err := client.UpdateTopic(ctx, topicUpdate1); err != nil {
+		t.Errorf("Failed to update topic: %v", err)
+	} else if !testutil.Equal(gotTopicConfig, wantUpdatedTopicConfig1) {
+		t.Errorf("UpdateTopic() got: %v, want: %v", gotTopicConfig, wantUpdatedTopicConfig1)
+	}
+
+	topicUpdate2 := &TopicConfigToUpdate{
+		Name:              topicPath,
+		PerPartitionBytes: 35 * gibi,
+		RetentionDuration: InfiniteRetention,
```

Comment:
Ack, we'll definitely need to expand on doc.go eventually.

Reply:
On the subject of infinite retention, I would appreciate a second opinion on something. More context in b/170457893#comment17.

So for topic creation, I treated TopicConfig.RetentionDuration = {InfiniteRetention|nil} to behave exactly the same, i.e. infinite retention. For topic update, TopicConfigToUpdate.RetentionDuration = nil means leave the retention as-is; = InfiniteRetention means to clear the current retention value and make it infinite.

WDYT? Is this confusing?

Alternatively, TopicConfig.RetentionDuration could be changed to a time.Duration, where an unset duration proto is represented as InfiniteRetention, rather than nil. This makes the semantics identical in TopicConfig and TopicConfigToUpdate.

Reply:
> Alternatively, TopicConfig.RetentionDuration could be changed to a time.Duration, where an unset duration proto is represented as InfiniteRetention, rather than nil. This makes the semantics identical in TopicConfig and TopicConfigToUpdate.

I just went ahead and implemented this, as I think having the same semantics may be less confusing. Can change it back if anyone likes the original implementation.